interfaces/builtin/overmount.go

```diff
@@ -0,0 +1,136 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+// OvermountInterface allows customizing the userspace exposed to the snap
+type OvermountInterface struct{}
+
+func (iface *OvermountInterface) Name() string {
+	return "overmount"
+}
+
+func (iface *OvermountInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	return nil
+}
+
+func (iface *OvermountInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	// check that we both a source and a destination
+	source, destination, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return err
+	}
+	if len(source) == 0 || len(destination) == 0 {
+		return fmt.Errorf("source and destination must be set")
+	}
+
+	return nil
+}
+
+func (iface *OvermountInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *OvermountInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *OvermountInterface) overmountMountSnippet(plug *interfaces.Plug, src, dst string) ([]byte, error) {
+	snippet := bytes.NewBuffer(nil)
+	fmt.Fprintln(snippet, overmountEntry(plug, src, dst, ",ro,exec"))
+	return snippet.Bytes(), nil
+}
+
+// Obtain yaml-specified source and destination to mount
+func (iface *OvermountInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
```

Comment:
s/getAttribs/mounts/; this will need to be tuned so it returns pairs of source/target.